x/aswap/handler.go

```diff
@@ -0,0 +1,292 @@
+	if err != nil {
+		return nil, err
+	}
+
+	res := &weave.CheckResult{
+		GasAllocated: createSwapCost,
+	}
+	return res, nil
+}
+
+// Deliver moves the tokens from sender to the swap account if all conditions are met.
+func (h CreateSwapHandler) Deliver(ctx weave.Context, db weave.KVStore, tx weave.Tx) (*weave.DeliverResult, error) {
+	msg, err := h.validate(ctx, db, tx)
+	if err != nil {
+		return nil, err
+	}
+
+	// apply a default for sender
+	sender := msg.Src
+	if sender == nil {
+		sender = x.MainSigner(ctx, h.auth).Address()
+	}
+
+	// create an swap object
+	swap := &Swap{
+		Metadata:     &weave.Metadata{},
+		Src:          sender,
+		PreimageHash: msg.PreimageHash,
+		Recipient:    msg.Recipient,
+		Timeout:      msg.Timeout,
+		Memo:         msg.Memo,
+	}
+	obj, err := h.bucket.Build(db, swap)
+	if err != nil {
+		return nil, err
+	}
+
+	// deposit amounts
+	senderAddr := swap.Src
+	if err := moveCoins(db, h.bank, senderAddr, weave.NewAddress(swap.PreimageHash), msg.Amount); err != nil {
+		return nil, err
+	}
+	// return id of swap to use in future calls
+	res := &weave.DeliverResult{
+		Data: obj.Key(),
+	}
+	return res, h.bucket.Save(db, obj)
+}
+
+// validate does all common pre-processing between Check and Deliver.
+func (h CreateSwapHandler) validate(ctx weave.Context, db weave.KVStore, tx weave.Tx) (*CreateSwapMsg, error) {
+	var msg CreateSwapMsg
+	if err := weave.LoadMsg(tx, &msg); err != nil {
+		return nil, errors.Wrap(err, "load msg")
+	}
+
+	if IsExpired(ctx, msg.Timeout) {
+		return nil, errors.Wrap(errors.ErrInvalidInput, "timeout in the past")
+	}
+
+	// Sender must authorize this
+	if !h.auth.HasAddress(ctx, msg.Src) {
+		return nil, errors.ErrUnauthorized
+	}
+
+	// Leave the most expensive operation till we've sanity-checked everything else.
+	_, err := loadSwap(h.bucket, db, msg.PreimageHash)
+	switch {
+	case !errors.ErrEmpty.Is(err):
+		return nil, err
+	case err == nil:
+		return nil, errors.Wrap(errors.ErrDuplicate, "swap with the same preimage")
+	}
+
+	return &msg, nil
+}
+
+// ReleaseSwapHandler releases the amount to recipient.
+type ReleaseSwapHandler struct {
+	auth   x.Authenticator
+	bucket Bucket
+	bank   cash.Controller
+}
+
+var _ weave.Handler = ReleaseSwapHandler{}
+
+// Check just verifies it is properly formed and returns
+// the cost of executing it
+func (h ReleaseSwapHandler) Check(ctx weave.Context, db weave.KVStore, tx weave.Tx) (*weave.CheckResult, error) {
+	_, err := h.validate(ctx, db, tx)
+	if err != nil {
+		return nil, err
+	}
+
+	return &weave.CheckResult{GasAllocated: releaseSwapCost}, nil
+}
+
+// Deliver moves the tokens from swap account to the receiver if
+// all preconditions are met. When the swap account is empty it is deleted.
+func (h ReleaseSwapHandler) Deliver(ctx weave.Context, db weave.KVStore, tx weave.Tx) (*weave.DeliverResult, error) {
+	swap, err := h.validate(ctx, db, tx)
+	if err != nil {
+		return nil, err
+	}
+
+	swapAddr := weave.NewAddress(swap.PreimageHash)
```

Comment:
pp: would love to have the address persisted and taken from the swap instead of recalculating it. There is no guarantee that we never touch address generation nor format.